src/eth/TransactionManager.js

```diff
@@ -2,19 +2,57 @@ import PublicService from '../core/PublicService';
 import TransactionObject from './TransactionObject';
 import { Contract } from 'ethers';
 import { dappHub } from '../../contracts/abis';
+import { uniqueId } from '../utils';
+import { has } from 'lodash';
+import debug from 'debug';
+const log = debug('dai:testing');
 
 let txId = 1;
 
+// decorator definition
+export function tracksTransactions(target, name, descriptor) {
+  const original = descriptor.value;
+  descriptor.value = function(...args) {
+    const lastArg = args[args.length - 1];
+    let options;
+    if (typeof lastArg === 'object' && lastArg.constructor === Object) {
+      args = args.slice(0, args.length - 1);
+      options = lastArg;
+    } else {
+      options = {};
+    }
+
+    const promise = (async () => {
```

Comment:
we set up a promise by declaring and immediately calling an async function, and keep a reference to it